### ISSUE TITLE
Long boot with devices with ":" in path

### ISSUE THREAD

**Issue by osmano807** [NixOS/nixpkgs#249836]
### Describe the bug
NixOS on boot loops waiting for devices with wrong names.

According to https://github.com/NixOS/nixpkgs/blob/30733fceb445ed9ac1aa2528ceb2ea6238d88d4a/nixos/modules/system/boot/stage-1-init.sh#L91C4-L91C4 device paths are split by ":", but for some persistent block device names we have ":" in valid paths.

```
❯ ls /dev/disk/by-path
pci-0000:00:07.1-ata-2    pci-0000:00:10.0-scsi-0:0:0:0        pci-0000:00:10.0-scsi-0:0:0:0-part2
pci-0000:00:07.1-ata-2.0  pci-0000:00:10.0-scsi-0:0:0:0-part1  pci-0000:00:10.0-scsi-0:0:0:0-part3
```

The boot is stuck trying to load:
```
waiting for device /dev/disk/by-path/pci-0000 to appear....
waiting for device 00 to appear....
waiting for device 10.0-scsi-0 to appear....
waiting for device 0 to appear....
waiting for device 0 to appear....
waiting for device 0-part3 to appear....
```

### Steps To Reproduce
1. Install NixOS using /dev/disk/by-path devices
2. Get boot waiting loop

### Expected behavior
Boot without waiting for nonexistant devices

### Screenshots
![image](https://github.com/NixOS/nixpkgs/assets/321379/eb457655-8154-48ec-a556-4043a346e3a0)

### Notify maintainers
@Madouura 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
╰─❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.39, NixOS, 23.05 (Stoat), 23.05.20230817.8fe0514`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - nixpkgs: `/nix/store/rwg22rlfgk967c9grlpba44m3ydrk8iy-source`
```

**Comment by Madouura** [comment]
This is a recent phenomenon I've experienced a few times.
I'm not entirely sure why this happens, but since Linux 6.4.8 I believe, I haven't had this issue.
There are two reasons why this could happen AFAIK:
1. Something is messed up between the linux kernel and stage-1.
2. udev needs to have all the devices loaded before trying to mount multiple.

That's the best I have ATM.

Current nix-info for reference:
 - system: `"x86_64-linux"`
 - host os: `Linux 6.4.10, NixOS, 23.05 (Stoat), 23.05.20230815.6bd56dd`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.5`
 - channels(mado): `"nixos-23.05, nixos-unstable"`
 - channels(root): `"nixos-23.05, nixos-unstable"`
 - nixpkgs: `/etc/nix/path/nixpkgs`

**Comment by Madouura** [comment]
I should also clarify in my case that it was a breaking bug, where I would have to reboot because one of the devices never loaded, until it worked.

**Comment by delroth** [comment]
@Madouura it seems like you're describing a different issue with bcachefs and multiple devices separated by `:`. This bug, as far as I understand, is about non-bcachefs block device paths that contain `:` incidentally (not as a multi-device delimiter). I'd suggest filing a separate issue for your problem.

**Comment by delroth** [comment]
(But as the author of the change that introduced that `:` behavior - #153095 - I suspect you might want to be involved here regardless :P)

**Comment by Madouura** [comment]
My bad, the usual bad reading comprehension and I didn't look at the paths that were listed closely enough.
I would suggest, if possible, using something like ``/dev/disk/by-uuid/XXXX`` instead until this is fixed.

As for the fix, while it would break the use-case of colon-separated ``/dev/disk/by-path`` entries, we could add a conditional to skip specifying the ``IFS`` if it is the aforementioned path.

**Comment by pjones** [comment]
This should be considered a bug.

There are *a lot* of paths in `/dev/disk/*` that have colons in their name (including RAID UUID devices).  Requiring users to know that devices with colons in their name are treated special is unreasonable, especially if it's not documented.

If bcachefs requires this then I think we need a new configuration option or the stage 1 `waitDevice` function needs to be extended so it considers the file system type before setting `IFS`.

@Madouura @Mic92

**Comment by Madouura** [comment]
IMO, we should get rid of it and just use the new ``UUID=`` mechanism.

**Comment by pjones** [comment]
@Madouura I don't use bcachefs so I don't know what the `UUID=` mechanism is.

Is it safe to simply revert your patch to `waitDevice`?

**Comment by Madouura** [comment]
> Is it safe to simply revert your patch to `waitDevice`?

Unfortunately no, because many people's auto-generated configs still use the ':' separator.
It's something I think we'll have to do at some point though.
I think the better option may be to omit a warning if it's being used (i.e: using bcachefs) until 24.05, backport that to 23.11, and remove it in 24.05.

**Comment by pjones** [comment]
I would suggest that we change the `waitDevice` function to take an optional second argument, the file system type.  If that argument is set *and* it is set to "`bcachefs`" then (and only then) do we set `IFS`.

Of course, this only works if the auto-generated configs correctly set the file system type.  @Madouura Can you point me to the commit where you changed the auto-generation process?

**Comment by Madouura** [comment]
Should be fixed in #279852
